hackathon_site/hardware/serializers.py

```diff
@@ -46,9 +54,128 @@ def get_unique_hardware_count(obj: Category):
         return Hardware.objects.filter(categories__id=obj.id).count()
 
 
-class OrderSerializer(serializers.ModelSerializer):
+class OrderListSerializer(serializers.ModelSerializer):
     hardware_set = HardwareSerializer(many=True, read_only=True)
 
     class Meta:
         model = Order
         fields = ("id", "hardware_set", "team", "status", "created_at", "updated_at")
+
+
+class OrderCreateSerializer(serializers.Serializer):
+    class OrderCreateHardwareSerializer(serializers.Serializer):
+        id = serializers.PrimaryKeyRelatedField(
+            queryset=Hardware.objects.all(), many=False, required=True
+        )
+        quantity = serializers.IntegerField(required=True)
+
+    hardware = OrderCreateHardwareSerializer(many=True, required=True)
+
+    @staticmethod
+    def merge_requests(hardware_requests):
+        return functools.reduce(
+            lambda x, y: x + y,
+            [Counter({e["id"]: e["quantity"]}) for e in hardware_requests],
+        )
+
+    # check that the requests are within per-team constraints
+    def validate(self, data):
+        requested_hardware = self.merge_requests(hardware_requests=data["hardware"])
+        hardware_query = (
+            Hardware.objects.filter(
+                id__in=[hardware.id for hardware in requested_hardware.keys()],
+            )
+            .all()
+            .prefetch_related("categories", "order_items")
+        )
+        team_unreturned_orders = hardware_query.annotate(
+            past_order_count=Count(
+                "order_items",
+                filter=Q(order_items__part_returned_health__isnull=True)
+                & ~Q(order_items__order__status="Cancelled")
+                & Q(order_items__order__team=self.context["request"].user.profile.team),
+            )
+        )
+        category_counts = dict()
+        error_messages = []
+        for (hardware, requested_quantity) in requested_hardware.items():
+            try:
+                team_hardware = team_unreturned_orders.get(id=hardware.id)
+            except ObjectDoesNotExist:
```

Comment:
The only time this would happens is if the id they submitted is invalid, right? I think in that case we should just throw a validation error since its a bad request, rather than swallowing it and moving on.

Reply:
it honestly shouldn't ever happen, since the hardware id is validated through the serializer field. but yea, not swallowing the error is a good idea.